src/protocols/core.rs

```diff
@@ -268,80 +327,31 @@ unsafe fn handle_create_vcpu_request(vmsa: *mut Vmsa) {
         Ok(v) => v,
         Err(_e) => return,
     };
-    let create_vmsa: *mut Vmsa = create_vmsa_va.as_mut_ptr();
 
     let vmpl: VMPL = VMPL::Vmpl1;
-    'main: loop {
-        // Revoke access to all non-zero VMPL levels to prevent tampering
-        // before checking the fields within the new VMSA.
-        let ret: u32 = revoke_vmpl_access(create_vmsa_va, RMP_4K);
-        if ret != 0 {
-            break;
+    let ret: u64 = match create_vcpu_inner(
+        apic_id,
+        vmpl,
+        create_vmsa_gpa,
+        create_vmsa_va,
+        create_ca_gpa,
+    ) {
+        Ok(()) => SVSM_SUCCESS,
+        Err(code) => {
+            // On error turn the page (back) into a non-VMSA page
+            grant_vmpl_access(create_vmsa_va, RMP_4K, vmpl as u8);
+            code
```

Comment:
And this still needs to go through the tlb flush instead of returning directly.

Reply:
It should go all the way to the end; this `match` statement just sets the `ret` variable, which is then used at the end `(*vmsa).set_rax(ret);`.

Reply:
Ah, my bad...  confused it with the implicit return thing. One of the things about Rust that I'm not crazy about, you don't always have to use "return" and so I misread this and forgot it is in a match statement.